### ISSUE TITLE
[Console] Progress bar is by default  printed to stderr, changing output is complicated

### ISSUE THREAD

**Issue by zdenekdrahos** [symfony/symfony#18744]
Progressbar is by default redirected to stderr. 
That's fine, if I could simple change output to `stdout`.

## Example

Given errors from cron are logged to file:

```
* * * * *       app/console something 1> /dev/null 2>> app/logs/cron.log
```

After upgrade to symfony 2.8 the file contains:

```
$ cat app/logs/cron.log

 0/3 [>---------------------------]   0%
 1/3 [=========>------------------]  33%
 2/3 [==================>---------]  66%
 3/3 [============================] 100%
```

## Why?

https://github.com/symfony/symfony/pull/15795 sets `getErrorOutput` in constructor. Provided solution:

> If someone explicitly wants to use stdout, they can simply pass $output->getStream() instead of $output in most use-cases.

doesn't work because:

```
  [Symfony\Component\Debug\Exception\ContextErrorException]
  Catchable Fatal Error: Argument 1 passed to Symfony\Component\Console\Helper\ProgressBar::__construct() must implement interface Symfony\Component\Console\Output\OutputInterface, resource given, called in /Users/www/ter_004_gosms/src/Analytics/Command/SmsReportingC
  ommand.php on line 41 and defined
```

## Solution

Temporarily I solved it with creating another output:

```
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $stdout = new \Symfony\Component\Console\Output\StreamOutput($output->getStream());
        $progressBar = new ProgressBar($stdout);
    }
```

That's not very good developer experience. I had to search github issues, PR, symfony/console code...
Printing progress to `stdout` should be at least mentioned in [documentation](http://symfony.com/doc/current/components/console/helpers/progressbar.html)

**Comment by Tobion** [comment]
Two lines of code doesn't look very complicated. If you think that needs to be covered by documentation, please open an issue on the symfony-docs repo. Thanks

**Comment by zdenekdrahos** [comment]
Fair enough.

I've created little helper for creating progress bars. Because it got little complicated when `NullOutput` doesn't have `getStream` method and progress bar is created in multiple places. Maybe someone will find it useful:

```php
<?php

namespace AppBundle\Command;

use Symfony\Component\Console\Output\StreamOutput;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Helper\ProgressBar;

class ProgressBarFactory
{
    /**
     * Avoid printing progress bar to stderr
     * https://github.com/symfony/symfony/issues/18744
     */
    public static function standardOutput(OutputInterface $output)
    {
        $stdout = $output instanceof StreamOutput ? new StreamOutput($output->getStream()) : $output;
        return new ProgressBar($stdout);
    }
}

```

**Comment by escopecz** [comment]
@zdenekdrahos thanks for your solution! I built my code on top of it, but I'm getting
```
PHP Fatal error:  Call to a member function getVerbosity() on null in vendor/symfony/console/Helper/ProgressBar.php on line 429
```
when `\Symfony\Component\Console\Output\StreamOutput` is used. Did you find solution to that?

@Tobion Why would anyone want to render output of ProgressBar as stderr? I can't even extend the ProgressBar class, because it's full of `protected` properties and even methods. Symfony 2.8 broke the BC with this.

**Comment by zdenekdrahos** [comment]
I didn't get `getVerbosity` error. I don't think that it is caused by `StreamOutput`. Every output implements [this method](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Console/Output/OutputInterface.php#L62), so [$this->output should not be null](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Console/Helper/ProgressBar.php#L58) in pure `ProgressBar`.

**Comment by pulzarraider** [comment]
@Tobion this behavior `ProgressBar` is absolutly nonstandard. In `stderr` output should be only the errors.
We are loging differently the `stderr` and `stdout` and sending output of ProgressBar in the `stderr` breaks it.

It's not good, if the developers are forced to fix the nonstandard behaviour of the framewrok if they want the standard. The default should be: `stdout`. `stderr` should be just the option.

**Comment by stof** [comment]
@pulzarraider no, stderr is not about rendering errors (Bash programming does not call it stderr btw). The second pipe is about displaying any informational message, while the first pipe (named stdout in PHP) is about displaying the main output of the command (anything you would pipe to another tool in a chain).
The issue is that many people are thinking that stderr is about displaying errors only, while this is not what it is about.

**Comment by escopecz** [comment]
@stof When we take in consideration that Symfony is a PHP framework, not a Bash script, it's maybe better to call it stderr as PHP does. But that's not the point. The point is:

1. The pipe was changed in minor version update which broke the backward compatibility and therefore is against [Semantic Versioning](http://semver.org/) which Symfony [should be following].(http://symfony.com/doc/current/contributing/code/bc.html).
2. When it was changed, there should be an easy way to configure it back. The proposed way doesn't work. 
3. Symfony maintainers doesn't care about 1. nor 2. and say: "Two lines of code doesn't look very complicated". But it's not 2 lines. It's quite ugly hack.

**Comment by javiereguiluz** [comment]
@escopecz Symfony is a PHP framework ... but here were are discussing about console commands, so it's like a Bash script.

In any case, the problem here is that both of you are correct. @stof is 100% right in his explanation (stderr is not only for errors) and @escopecz is 100% right too because, in the real world, everybody is doing this wrong and considers stderr = errors only.

**Comment by pulzarraider** [comment]
@stof I accept your argument, that 
I agree with @escopecz. This is a BC break that should not be merged in the `2.7`+. Should be marked as feature and released in new version, marked in the UPGRDAE docs, so we can prepare our codes for this change.

**Comment by fabpot** [comment]
I agree that it should not have been merged in the 2.7 branch. But back then, we believed it should be considered as a bug and we did not anticipate that it would cause any harm. I'm very sorry about that. Just to add on the topic of stderr, nobody should ever rely on errors being sent to stderr. A very good example being git; git sends a lot of things on stderr that are not error messages. And having the progress bar on stdout means that redirecting the output to a file means getting the progress bar as well, which we consider as a bug.
